src/pal/src/thread/process.cpp

```diff
@@ -1620,52 +1580,75 @@ class PAL_RuntimeStartupHelper
         }
 
     exit:
-        if (*pCoreClrExists)
-        {
-            // Wake up all the runtimes
-            if (sem_post(m_continueSem) != 0)
-            {
-                ASSERT("sem_post(continueSem) failed: errno is %d (%s)\n", errno, strerror(errno));
-            }
-        }
-
         if (listHead != NULL)
         {
             DestroyProcessModules(listHead);
         }
+
         return pe;
     }
 
     void StartupHelperThread()
     {
-        bool coreclrExists = false;
+        char continueSemName[NAME_MAX - 4];
+        sem_t *continueSem = SEM_FAILED;
+        PAL_ERROR pe = NO_ERROR;
+
+        sprintf_s(continueSemName,
+                  sizeof(continueSemName),
+                  RuntimeContinueSemaphoreName,
+                  m_processId,
+                  m_processIdDisambiguationKey);
 
-        PAL_ERROR pe = InvokeStartupCallback(&coreclrExists);
-        if (pe == NO_ERROR)
+        TRACE("StartupHelperThread continue sem '%s'\n", continueSemName);
+
+        // Does the continue semaphore exists? If it does, the runtime is ready to be debugged.
+        continueSem = sem_open(continueSemName, O_RDWR);
+        if (continueSem != SEM_FAILED)
```

Comment:
I could check errno == ENOENT and return any other error (via the callback).  It is probably pretty important for the debugger to “see” the EACCES errors.

Return Value
On success, sem_open() returns the address of the new semaphore; this address is used when calling other semaphore-related functions. On error, sem_open() returns SEM_FAILED, with errno set to indicate the error.
Errors
EACCES
The semaphore exists, but the caller does not have permission to open it.
EEXIST
Both O_CREAT and O_EXCL were specified in oflag, but a semaphore with this name already exists.
EINVAL
value was greater than SEM_VALUE_MAX.
EINVAL
name consists of just "/", followed by no other characters.
EMFILE
The process already has the maximum number of files and open.
ENAMETOOLONG
name was too long.
ENFILE
The system limit on the total number of open files has been reached.
ENOENT
The O_CREAT flag was not specified in oflag and no semaphore with this name exists; or, O_CREAT was specified, but name wasn't well formed.
ENOMEM
Insufficient memory.